README
Parachute mod for Minecraft (Forge version)
Authored by crackedEgg


ChangeLog
commit 529f0487d1ef1185be1b18df0f58808a991cf2ce
Author: Michael Sheppard <pbrane0@gmail.com>
Date:   Tue Nov 6 09:03:41 2018 -0500

    Added a flying sound, like the elytra
    added an option to toggle the flying sound
    changed logger instantiation to event.getModLog()

    Signed-off-by: Michael Sheppard <pbrane0@gmail.com>

commit 06565aa752e562b1bac21addeb344bcace3a7b00
Author: Michael Sheppard <pbrane0@gmail.com>
Date:   Wed Oct 31 15:50:22 2018 -0400

    added PlayerHurtEvent back in

    Signed-off-by: Michael Sheppard <pbrane0@gmail.com>

commit eead69a808d3d49f6066ea598d2e0a3253c7fbdd
Author: Michael Sheppard <pbrane0@gmail.com>
Date:   Tue Oct 30 18:15:17 2018 -0400

    removed hud-bubble option
    removed altitudeMSL option

    Signed-off-by: Michael Sheppard <pbrane0@gmail.com>

commit dbecb42a28edc6521a45ab7ca09bd5f17faec355
Author: Michael Sheppard <pbrane0@gmail.com>
Date:   Tue Oct 30 16:32:32 2018 -0400

    comment out test for vines

    Signed-off-by: Michael Sheppard <pbrane0@gmail.com>

commit bc1f77b5848a4cb2f23490db487663fab0fa0d55
Author: Michael Sheppard <pbrane0@gmail.com>
Date:   Tue Oct 30 11:47:52 2018 -0400

    added keybinding to toggle HUD visibility
    removed noHUD config option and all associated code

    Signed-off-by: Michael Sheppard <pbrane0@gmail.com>

commit 178f930d81977bad0deb212fbf520848c338fb06
Author: Michael Sheppard <pbrane0@gmail.com>
Date:   Mon Oct 29 19:06:33 2018 -0400

    fixed client/server aadActive state
    removed aad altitude and immediate options
    * aad is now immediate only, 5.0 block minimum fall distance

    Signed-off-by: Michael Sheppard <pbrane0@gmail.com>

commit bad1a3230b51cd351a327f87384d936c19e068a2
Author: Michael Sheppard <pbrane0@gmail.com>
Date:   Mon Oct 29 13:04:06 2018 -0400

    added network packet for AAD state between client/server

    Signed-off-by: Michael Sheppard <pbrane0@gmail.com>

commit 2399dbaa80c6b52c0d1022bbac17b976e1d43042
Author: Michael Sheppard <pbrane0@gmail.com>
Date:   Sun Oct 28 17:19:01 2018 -0400

    removed playerfallevent, playerhurtevent, and playermountevent
    and associated
    code in entityparachute
    removed event registering code in main mod class

    Signed-off-by: Michael Sheppard <pbrane0@gmail.com>

commit 8a9f067916fb0115e98e2530fa761fa6bbb2203a
Author: Michael Sheppard <pbrane0@gmail.com>
Date:   Sun Oct 28 16:12:59 2018 -0400

    fixed player not colliding with ground or other objects
    fixed PlayerLoginHandler not adding player to list

    Signed-off-by: Michael Sheppard <pbrane0@gmail.com>

commit 4bbbfc540a2662a8f6441ede242c6923de3623b8
Author: Michael Sheppard <pbrane0@gmail.com>
Date:   Wed Jun 6 16:41:29 2018 -0400

    changed reticule for no bubble case.

    Signed-off-by: Michael Sheppard <pbrane0@gmail.com>

commit f76cd36a7e84927872eb35de323e420a39a6ccbe
Author: m1k3s <pbrane0@gmail.com>
Date:   Mon Jun 4 08:30:51 2018 -0400

    modified HUD image, now user can hide bubble
    added final SCALE variable

    Signed-off-by: m1k3s <pbrane0@gmail.com>

commit a5074e8380ccf9c3aa2491f768fb93d25d2dddb3
Author: m1k3s <pbrane0@gmail.com>
Date:   Thu May 17 10:27:29 2018 -0400

    cleaned up some code, reformatted
    added config option to disable display of the player front facing
    bubble on the HUD
    removed args from methods in HUD renderer, made those args class
    fields

    Signed-off-by: m1k3s <pbrane0@gmail.com>

commit b457368227bf351e494993755bb39189910494df
Author: m1k3s <pbrane0@gmail.com>
Date:   Sat Apr 28 15:36:13 2018 -0400

    adjusted setSize to correct dimensions
    updated ChangeLog

    Signed-off-by: m1k3s <pbrane0@gmail.com>

commit b1ee7567a98e76b2b4f6fbc927998fc2f307586f
Author: m1k3s <pbrane0@gmail.com>
Date:   Mon Apr 16 07:51:28 2018 -0400

    tweaking the new parachute, six sectios, thicker
    changed the tracking update interval to 5

    Signed-off-by: m1k3s <pbrane0@gmail.com>

commit 3370625939efbcc9f939f5c7094a6b1a02ee2e9f
Author: m1k3s <pbrane0@gmail.com>
Date:   Mon Apr 9 14:50:18 2018 -0400

    larger canopy, six sections

    Signed-off-by: m1k3s <pbrane0@gmail.com>

commit 877b91a4d7aa1eb371628358a1d5ff8b960edcaa
Author: m1k3s <pbrane0@gmail.com>
Date:   Sun Mar 25 11:00:39 2018 -0400

    remove unused initialVelocity variable, merge residual
    code clean up

    Signed-off-by: m1k3s <pbrane0@gmail.com>

commit 0801273f4be73f0ce96138606683e188d0730849
Merge: 7e458de 7251527
Author: m1k3s <pbrane0@gmail.com>
Date:   Sun Mar 18 11:57:34 2018 -0400

    Merge branch 'master' into 1.12.2

    # Conflicts:
    #       parachute/client/ParachuteInputEvent.java
    #       parachute/common/ConfigHandler.java
    #       parachute/common/EntityParachute.java

    Signed-off-by: m1k3s <pbrane0@gmail.com>

commit 72515271ee93414e17cff5ad000159bf7c6d9503
Merge: b3261a5 2c3c2f2
Author: m1k3s <pbrane0@gmail.com>
Date:   Sun Mar 18 11:45:12 2018 -0400

    Merge branch 'newhud'

commit 2c3c2f28244238cf72d87e6b559bc3b003d6b65b
Author: m1k3s <pbrane0@gmail.com>
Date:   Sun Mar 18 11:44:52 2018 -0400

    made all finals uppercase
    changed some field names to better reflect their purpose
    reworked the steer by sight option as a toggle.

    Signed-off-by: m1k3s <pbrane0@gmail.com>

commit b3261a5abe4342c5d63569bd063d937f647cac08
Merge: a26c111 887c275
Author: m1k3s <pbrane0@gmail.com>
Date:   Tue Mar 13 15:52:08 2018 -0400

    Merge branch 'newhud'

commit 887c275a0a3896d0e81b16cd9febdb6217630bbe
Author: m1k3s <pbrane0@gmail.com>
Date:   Tue Mar 13 15:51:44 2018 -0400

    steerBySight is now a local variable in EntityParachute.java
    commented out version checking in Parachute.java

    Signed-off-by: m1k3s <pbrane0@gmail.com>

commit a26c111239047babf630f33334d1de24e5c5df7d
Merge: d796967 bc3cfb7
Author: m1k3s <pbrane0@gmail.com>
Date:   Mon Mar 12 18:19:50 2018 -0400

    Merge branch 'newhud'

commit bc3cfb7103d1e60d3c1a5823477b043c7eaa70c3
Author: m1k3s <pbrane0@gmail.com>

parachute/client/HudGuiRenderer.java
/*
 * HudGuiRenderer.java
 *
 * Copyright (c) 2017 Michael Sheppard
 *
 *  =====GPL=============================================================
 * $program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see http://www.gnu.org/licenses.
 * =====================================================================
 *
 */
package com.parachute.client;

import com.parachute.common.ConfigHandler;
import com.parachute.common.EntityParachute;
import com.parachute.common.Parachute;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.FontRenderer;
import net.minecraft.client.gui.Gui;
import net.minecraft.client.gui.ScaledResolution;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.MathHelper;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.client.event.RenderGameOverlayEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;


@SideOnly(Side.CLIENT)
public class HudGuiRenderer extends Gui {

    // the parachute-hud.png image uses some ELD Unofficial Continuation Project textures, modified to suit the parachute HUD.
    protected static final ResourceLocation hudTexture = new ResourceLocation(Parachute.MODID + ":" + "textures/gui/parachute-hud.png");
    private static FontRenderer fontRenderer;
    public static double altitude;
    private final Minecraft mc = Minecraft.getMinecraft();

    private int blink = 0;
    private final int hudWidth;
    private final int hudHeight;
    private final int ledWidth;
    private final int ledHeight;
    private final int fieldWidth;
    private final int colorYellow;
    private final int colorRed;
    private final int colorGreen;
    private final int colorBlue;
    private final int colorDimGreen;
    private final int colorDimRed;
    private final int colorDimYellow;
    // AAD icon
    private final int aadWidth;
    private final int aadHeight;
    private final int ledY;
    // manual dismount indicators
    private final int lightY;
    private final int red;
    private final int darkRed;
    private final int green;
    private int blinkX;
    private final int blinkTime;
    private final int yOffset;
    private final int bigLedXY;

    double compassHeading;

    public HudGuiRenderer() {
        super();
        hudWidth = 182;
        hudHeight = 39;
        ledWidth = 11;
        ledHeight = 5;
        colorYellow = 0xffaaaa00;
        colorDimYellow = 0xcc888800;
        colorRed = 0xffaa0000;
        colorDimRed = 0xcc880000;
        colorGreen = 0xff00aa00;
        colorBlue = 0xff0000aa;
        colorDimGreen = 0xcc008800;
        aadWidth = 16;
        aadHeight = 25;
        bigLedXY = 16;
        ledY = 39;
        lightY = 44;
        red = 0;
        darkRed = 48;
        green = 16;
        blinkX = red;
        blinkTime = 5;
        yOffset = 14;

        fontRenderer = mc.fontRenderer;
        fieldWidth = fontRenderer.getStringWidth("000.0") / 2;
    }

    @SuppressWarnings("unused")
    @SubscribeEvent
    public void drawCenterHUD(RenderGameOverlayEvent.Post event) {
        if (event.isCancelable() || mc.gameSettings.showDebugInfo || mc.player.onGround) {
            return;
        }
        if (ClientConfiguration.getNoHUD() || ClientConfiguration.getUseCompassHUD()) {
            return;
        }
        if (mc.inGameHasFocus && event.getType() == RenderGameOverlayEvent.ElementType.ALL) {
            ScaledResolution sr = new ScaledResolution(mc);
            int hudX = (sr.getScaledWidth() / 2) - (hudWidth / 2); // left edge of GUI
            int hudY = 2; // top edge of GUI
            int textX = hudX + 30; // xcoord for text
            int textY = hudY + 22; // ycoord for text

            if (mc.player.getRidingEntity() instanceof EntityParachute) {
                mc.getTextureManager().bindTexture(hudTexture);
                EntityParachute chute = (EntityParachute) mc.player.getRidingEntity();
                if (chute == null) {
                    return;
                }

                GlStateManager.enableRescaleNormal();
                GlStateManager.enableBlend();
                GlStateManager.tryBlendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA,
                        GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO);

                BlockPos entityPos = new BlockPos(mc.player.posX, mc.player.getEntityBoundingBox().minY, mc.player.posZ);
                altitude = getCurrentAltitude(entityPos);
                double homeDir = getHomeDirection(chute.rotationYaw);
                double distance = getHomeDistance();
                compassHeading = calcCompassHeading(chute.rotationYaw);

                // Params: int screenX, int screenY, int textureX, int textureY, int width, int height
                drawTexturedModalRect(hudX, hudY, 0, 0, hudWidth, hudHeight); // draw the main hud
                drawTexturedModalRect(hudX, hudY + hudHeight, 0, 60, hudWidth, 9); // draw the chute bubble frame

                // determine which LED to light, homeDir is in range -180 to 180
                // for any value under -80 or over 80 the LED is fixed to the
                // left or right end of the slider respectively.
                int ledX = calcHomeDirection(homeDir);
                drawTexturedModalRect(hudX + ledX, hudY, ledX, ledY, ledWidth, ledHeight); // draw the lit LED

                // this indicator points to the parachute facing direction. once the
                // red indicator is centered the player is facing in the same
                // direction as the chute.
                double playerLook = MathHelper.wrapDegrees(mc.player.getRotationYawHead() - chute.rotationYaw);
                int chuteBubble = calcPlayerChuteFacing(playerLook);
                drawTexturedModalRect(hudX + chuteBubble, hudY + hudHeight, 67, 46, 3, 7); // draw the chute bubble

                // AAD status
                int aadIconX;
                int aadIconY = 8;
                aadIconX = ConfigHandler.getIsAADActive() ? 199 : 182;
                drawTexturedModalRect(hudX + hudWidth + 2, hudY + 8, aadIconX, aadIconY, aadWidth, aadHeight); // draw the AAD indicator

                // ground proximity indicator
                if (altitude > 10) {
                    drawTexturedModalRect(hudX - 18, hudY + yOffset, green, lightY, bigLedXY, bigLedXY);
                } else if (altitude <= 10 && altitude > 3) {
                    drawTexturedModalRect(hudX - 18, hudY + yOffset, red, lightY, bigLedXY, bigLedXY);
                } else if (altitude <= 3) { // make this blink
                    if ((blink % blinkTime) == 0) {
                        blinkX = blinkX == red ? darkRed : red;
                    }
                    drawTexturedModalRect(hudX - 18, hudY + yOffset, blinkX, lightY, bigLedXY, bigLedXY);
                    blink++;
                }

                // draw the altitude text
                fontRenderer.drawStringWithShadow(format(altitude), (textX + 5) - fieldWidth, textY, colorAltitude());
                // draw the compass heading text
                fontRenderer.drawStringWithShadow(format(compassHeading), (textX + 118) - fieldWidth, textY, colorDimGreen);//colorCompass(compassHeading));
                // draw the distance to the home point text
                fontRenderer.drawStringWithShadow(format(distance), (textX + 65) - fieldWidth, textY, colorDimGreen);

                GlStateManager.disableRescaleNormal();
                GlStateManager.disableBlend();
            }
        }
    }

    public String format(double d) {
        return String.format("%.1f", d);
    }

    private double calcCompassHeading(double yaw) {
        return (((yaw + 180.0) % 360) + 360) % 360;
    }

    private int calcPlayerChuteFacing(double playerLook) {
        int bubble = (int) Math.floor(playerLook + 91.0);
        bubble = bubble < 0 ? 0 : bubble > 179 ? 179 : bubble;
        return bubble;
    }

    private int calcHomeDirection(double dir) {
        int ledX = (int) Math.floor(dir + 84.0);
        ledX = ledX < 0 ? 0 : ledX > 170 ? 170 : ledX;
        return ledX;
    }

    // difference angle in degrees the chute is facing from the home point.
    // zero degrees means the chute is facing the home point.
    // the home point can be either the world spawn point or a waypoint
    // set by the player in the config.
    public double getHomeDirection(double yaw) {
        BlockPos blockpos = mc.world.getSpawnPoint();
        double delta = Math.atan2(blockpos.getZ() - mc.player.posZ, blockpos.getX() - mc.player.posX);
        double relAngle = delta - Math.toRadians(yaw);
        return MathHelper.wrapDegrees(Math.toDegrees(relAngle) - 90.0); // degrees
    }

    public double getHomeDistance() {
        BlockPos blockpos = mc.world.getSpawnPoint();
        double a = Math.pow(blockpos.getZ() - mc.player.posZ, 2);
        double b = Math.pow(blockpos.getX() - mc.player.posX, 2);
        return Math.sqrt(a + b);
    }

    public int colorAltitude() {
        return (altitude <= 10.0 && altitude >= 0.0) ? colorDimRed : altitude <= 15.0 && altitude > 10.0 ? colorDimYellow : colorDimGreen;
    }

    // quadrant color code
    // 315 to 44 green, mostly north
    // 45 to 134 yellow, mostly east
    // 135 to 224 red, mostly south
    // 225 to 314 blue, mostly west
    @SuppressWarnings("unused")
    public int colorCompass(double d) {
        return (d >= 0 && d < 45.0) ? colorGreen : (d >= 45.0 && d < 135.0) ? colorYellow :
                (d >= 135.0 && d < 225.0) ? colorRed : (d >= 225.0 && d < 315.0) ? colorBlue : colorGreen;
    }

    // calculate altitude in meters above ground. starting at the entity
    // count down until a non-air block is encountered.
    // only allow altitude calculations in the surface world
    // return a weirdly random number if in nether or end.
    public double getCurrentAltitude(BlockPos entityPos) {
        if (mc.world.provider.isSurfaceWorld()) {
            BlockPos blockPos = new BlockPos(entityPos.getX(), entityPos.getY(), entityPos.getZ());
            while (mc.world.isAirBlock(blockPos.down())) {
                blockPos = blockPos.down();
            }
            // calculate the entity's current altitude above the ground
            return entityPos.getY() - blockPos.getY();
        }
        return 1000.0 * mc.world.rand.nextGaussian();
    }

}
